src/api-state/reducer.js

```diff
@@ -0,0 +1,126 @@
+// @flow
+import {
+	API_REQ,
+	API_RESP_SUCCESS,
+	API_RESP_ERROR,
+	API_RESP_FAIL,
+	API_RESP_COMPLETE,
+} from './sync/apiActionCreators';
+import { CACHE_SUCCESS } from './cache/cacheActionCreators';
+
+type ApiState = {
+	[string]: QueryResponse,
+	inFlight: Array<string>,
+	fail?: boolean,
+};
+
+export const DEFAULT_API_STATE: ApiState = { inFlight: [] };
+export const DEFAULT_APP_STATE = {};
+
+type ObjectFilter = (Object, ...args?: Array<any>) => Object;
+const filterKeys: ObjectFilter = (obj: ApiState, keys: Array<string>) =>
+	Object.keys(obj).reduce(
+		(newObj: ApiState, key: string) => {
+			if (!keys.includes(key)) {
+				newObj[key] = obj[key];
+			}
+			return newObj;
+		},
+		{ ...DEFAULT_API_STATE }
+	);
+
+export const responseToState = (
+	response: QueryResponse
+): { [string]: QueryResponse } => ({ [response.ref]: response });
+
+/*
+ * The primary reducer for data provided by the API
+ */
+export function api(
+	state: ApiState = DEFAULT_API_STATE,
+	action: FluxStandardAction
+): ApiState {
+	switch (action.type) {
+		case API_REQ: {
+			const requestRefs = (action.payload || []).map(({ ref }) => ref);
+			const inFlight = state.inFlight
+				.filter(ref => !requestRefs.includes[ref]) // clean out current duplicates
+				.concat(requestRefs); // add new requested refs
+
+			if ((action.meta || {}).logout) {
+				// clear app state during logout
+				return { ...DEFAULT_API_STATE, inFlight };
+			}
+
+			// remove any `ref`s that are being refreshed - eliminate stale data
+			const newState = {
+				...filterKeys(state, requestRefs),
+				inFlight,
+			};
+
+			return newState;
+		}
+		case API_RESP_SUCCESS: // fall though
+		case API_RESP_ERROR:
+		case CACHE_SUCCESS: // fall through
+			// each of these actions provides an API response that should go into app
+			// state - error responses will contain error info
+			delete state.fail; // if there are any values, the API is not failing
+			return {
+				...state,
+				...responseToState((action.payload || {}).response),
+			};
+		case API_RESP_FAIL:
```

Comment:
since we're batching requests to the API, we need a way of distinguishing a failed batch request from individual failed API requests - i.e. if the Node API proxy throws an error, it's a 'failure' that needs to be handled at the root application level, whereas REST API response errors should generally be dealt with by the particular requesting container.

This is something that I expect we'll need to build a generic toast/UI for sometime soon - it essentially means the app is broken and probably can't recover (although it may just be a temporary network issue)